### PR TITLE
Implement QuicFileUtils

### PR DESCRIPTION
**Pull request by danzh2010**
Add QuicFileUtilsImpl using Envoy::FileSystem.

Risk Level: low
Testing: Added tests in test/extensions/quic_listeners/quiche/platform/quic_platform_test.cc and tested with --define quiche=enabled
Part of #2557

### CODE REVIEW

File: test/extensions/quic_listeners/quiche/platform/quic_platform_test.cc [hunk 1]
```diff
@@ -452,5 +457,65 @@ TEST(QuicPlatformTest, QuicTestOutput) {
...
+
+TEST_F(FileUtilsTest, ReadDirContents) {
+  addSubDirs({"sub_dir1", "sub_dir2", "sub_dir1/sub_dir1_1"});
+  addFiles({"file", "sub_dir1/sub_file1", "sub_dir1/sub_dir1_1/sub_file1_1", "sub_dir2/sub_file2"});
+
+  const std::unordered_set<std::string> expected = {
+      dir_path_ + "/file", dir_path_ + "/sub_dir1/sub_file1",
+      dir_path_ + "/sub_dir1/sub_dir1_1/sub_file1_1", dir_path_ + "/sub_dir2/sub_file2"};
+  std::vector<std::string> files = ReadFileContents(dir_path_);
+  EXPECT_EQ(expected.size(), files.size());
```
Comment (wu-bin) [inline test/extensions/quic_listeners/quiche/platform/quic_platform_test.cc]: Line 505 to 508: You can use EXPECT_THAT(files, UnorderedElementsAre(expected))

Comment (danzh2010) [reply test/extensions/quic_listeners/quiche/platform/quic_platform_test.cc]: good point!

File: source/extensions/quic_listeners/quiche/platform/quic_file_utils_impl.cc [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+void DFTraverseDirectory(const std::string& dirname, std::vector<std::string>& files) {
+  Envoy::Filesystem::Directory directory(dirname);
+  for (const Envoy::Filesystem::DirectoryEntry& entry : directory) {
+    switch (entry.type_) {
+    case Envoy::Filesystem::FileType::Regular:
+      files.push_back(absl::StrCat(dirname, "/", entry.name_));
+      break;
+    case Envoy::Filesystem::FileType::Directory:
+      if (entry.name_ != "." && entry.name_ != "..") {
+        DFTraverseDirectory(absl::StrCat(dirname, "/", entry.name_), files);
```
[the remark this answers is not in the document]
Comment (danzh2010) [reply source/extensions/quic_listeners/quiche/platform/quic_file_utils_impl.cc]: Yes, but the file name returned also contains double "/". The only place used in quic gives path with no-trailing '/'. I added comment that it takes path without trailing '/'

[the remark this answers is not in the document]
Comment (wu-bin) [reply source/extensions/quic_listeners/quiche/platform/quic_file_utils_impl.cc]: SG given how infrequent this is used. In general, I'd prefer ReadFileContentsImpl to remove the trailing '/' if needed.

File: source/extensions/quic_listeners/quiche/platform/quic_file_utils_impl.cc [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+#include "common/filesystem/directory.h"
+#include "common/filesystem/filesystem_impl.h"
+
+#include "absl/strings/str_cat.h"
+
+namespace quic {
+namespace {
+
+void DFTraverseDirectory(const std::string& dirname, std::vector<std::string>& files) {
```
Comment (alyssawilk) [inline source/extensions/quic_listeners/quiche/platform/quic_file_utils_impl.cc]: Naming?  I'm not entirely sure what DF stands for here.

File: source/extensions/quic_listeners/quiche/platform/quic_file_utils_impl.cc [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+      ASSERT(false,
+             absl::StrCat("Unknow file entry type ", entry.type_, " under directory ", dirname));
+    }
+  }
+}
+
+} // namespace
+
+// Traverses the directory |dirname| and returns all of the files it contains.
+std::vector<std::string> ReadFileContentsImpl(const std::string& dirname) {
```
Comment (alyssawilk) [inline source/extensions/quic_listeners/quiche/platform/quic_file_utils_impl.cc]: Comment for "upstream", I think this is badly named as it's not reading the file contents.
I don't think you can fix that here though :-/

File: source/extensions/quic_listeners/quiche/platform/quic_file_utils_impl.cc (hunk 3 above)
Comment (danzh2010) [reply source/extensions/quic_listeners/quiche/platform/quic_file_utils_impl.cc]: Rename to depthFirstTraverseDirectory

**Pull request merged**